tools/autograd/gen_python_functions.py

```diff
@@ -483,130 +452,224 @@ def is_noarg_binding(overloads):
-        )
-    else:
-        check_has_torch_function = TORCH_FUNCTION_CHECK.substitute(
-            namespace="THPVariableClass",
-            modulename='"torch.Tensor"',
-            self_="self_" if is_python_method else "nullptr",
-        )
-
-    max_args = max([get_python_argc(decl) for decl in declarations])
-    traceable = 'true' if all(should_trace(d) for d in declarations) else 'false'
-
     return template.substitute(
         name=name,
         pycname=pycname,
         method_header=method_header,
-        max_args=max_args,
+        max_args=max(map(lambda o: o.signature.arguments_count(), overloads)),
         signatures=signatures,
         traceable=traceable,
-        check_has_torch_function=check_has_torch_function,
+        check_has_torch_function=gen_has_torch_function_check(
+            name=name,
+            module=module,
+            noarg=noarg,
+            method=method,
+        ),
         dispatch=dispatch,
         method_footer=method_footer,
-        self_="self_" if is_python_method else "nullptr",
+        self_="self_" if method else "nullptr",
     )
 
+def gen_has_torch_function_check(name: str, module: str, *, noarg: bool, method: bool) -> str:
+    if noarg:
+        return f"""\
+if(check_has_torch_function(self_)) {{
+  return handle_torch_function(self_, "{name}");
+}}
+""" if method else ''
+
+    self_ = "self_" if method else "nullptr"
+    namespace = {
+        "torch": "THPVariableFunctionsModule",
+        "torch.nn": "THPNNVariableFunctionsModule",
+        "torch.fft": "THPFFTVariableFunctionsModule",
+        "torch.linalg": "THPLinalgVariableFunctionsModule",
+    }[module] if module else "THPVariableClass"
+
+    return f"""\
+if(_r.has_torch_function()) {{
+  return handle_torch_function(_r, {self_}, args, kwargs, {namespace}, "{module or "torch.Tensor"}");
+}}
+"""
 
-#
-# forward declarations
-#
-
-PY_VARIABLE_FUNCTION_VARARGS_FORWARD_DECLARATION = CodeTemplate("""\
-static PyObject * ${pycname}(PyObject* self_, PyObject* args, PyObject* kwargs);
+# handler for output/no-output overload pair
+PY_VARIABLE_OUT = CodeTemplate("""\
+if (_r.isNone(${out_idx})) {
+  ${call_dispatch}
+} else {
+  ${call_dispatch_out}
+}
 """)
 
-PY_VARIABLE_FUNCTION_NOARGS_FORWARD_DECLARATION = CodeTemplate("""\
-static PyObject * ${pycname}(PyObject* self_, PyObject* args);
-""")
+def emit_dispatch_case(
+    overload: PythonSignatureGroup,
+    namedtuple_typenames: Dict[str, str],
+) -> str:
+    """
+    Emit dispatch code for a single parsed signature. This corresponds to either
+    a single native function, or a pair that differ only in output params. In the
+    latter case, a single python signature is used for both and dispatching
+    switches on the presence/absence of passed output args.
+    """
+    if overload.outplace is not None:
+        # dispatch output and no-output variants, branch on _r.isNone(<out_idx>)
+        return PY_VARIABLE_OUT.substitute(
```

Comment:
I noticed that the new codegen is moving away from CodeTemplate - but the auto-indent feature is useful in this case as the emit_single_dispatch can have different indents: inside switch-case, inside if-else, inside both or none.

Reply:
Yes, it is troublesome. For now, I have have been opting for code gen readability (f-strings) at the expense of generated code readability (yolo indentation) but generated code readability is getting worse and worse.  I'm kind of pissed that Python doesn't have some sort of pluggable f-string mechanism for this case.

I can only really think of a few options for this problem:
1. Run some sort of formatter (clang-format, maybe?) on the generated code after we generate it. Sounds slow.
2. Manually annotate multiline string formats with an indent level that triggers indentation, e.g., 

```
void f() {{
    {multiline(thing, 4)}
}}
```

3. Stop using f-strings